File: Labs/Lab4/Main.java
import java.util.Scanner;


public class Main {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) {
            return root;
        }
        if (root.left != null) {
            invertTree(root.left);
        }
        if (root.right != null) {
            invertTree(root.right);
        }
        TreeNode temp = root.left;
        root.left = root.right;
        root.right = temp;
        return root;
    }
    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        TreeNode node1 = new TreeNode(2);
        TreeNode node2 = new TreeNode(8);
        root.left = node1;
        root.right = node2;

        TreeNode node3 = new TreeNode(3);
        TreeNode node4 = new TreeNode(4);
        node1.left = node3;
        node1.right = node4;

        TreeNode node5 = new TreeNode(5);
        TreeNode node6 = new TreeNode(6);
        node2.left = node5;
        node5.right = node6;
        System.out.println(root.val);
        System.out.println(root.left.val + " " + root.right.val);
        System.out.println(root.left.left.val + " " + root.left.right.val + " " + root.right.left.val + " E" );
        System.out.println("E E E E E " + root.right.left.right.val);
        Main obj = new Main();
        root = obj.invertTree(root);
        System.out.println(root.val);
        System.out.println(root.left.val + " " + root.right.val);
        System.out.println("E " + root.left.right.val + " " + root.right.left.val + " " + root.right.right.val);
        System.out.println(" " + root.left.right.left.val + " E E E E E");
    }
}

File: Labs/Lab4/main.py
class TreeNode(object):

    def __init__(self, val=0, left=None, right=None):
        self.val = val

        self.left = left

        self.right = right


def invertTree(self, root):
    # if there is no root exit
    if not root:
        return root
    # if there is a left node go down the tree
    if root.left:
        invertTree(invertTree, root.left)
    # if there is a right node go down the tree
    if root.right:
        invertTree(invertTree, root.right)
    # swap the nodes at the bottom if one was empty then it still can swap
    temp = root.left
    root.left = root.right
    root.right = temp
    return root


# Creates test Tree
root = TreeNode(1)
node1 = TreeNode(2)
node2 = TreeNode(8)
root.left = node1
root.right = node2

node3 = TreeNode(3)
node4 = TreeNode(4)
node1.left = node3
node1.right = node4

node5 = TreeNode(5)
# node6 = TreeNode(6)
node2.left = node5
# node2.right = node6

# Print test tree
print(root.val)
print(root.left.val, root.right.val)
print(root.left.left.val, root.left.right.val, root.right.left.val)
# Invert test tree
invertTree(invertTree, root)
# Print inverted test tree
print(root.val)
print(root.left.val, root.right.val)
print("", root.left.right.val, root.right.left.val, root.right.right.val)
